File: sem iv/Theory of Computaion/lab/code/c/9_nfa_contains_001.cpp
// NFA (simulated deterministically) accepting all strings containing 001 as substring
// States: A=0 (start), B=1 (seen 0), C=2 (seen 00), D=3 (seen 001 -> accepting)
#include <iostream>
using namespace std;

int NFA_substring(string w) {
    int current = 0;
    for (char c : w) {
        switch (current) {
            case 0: current = (c == '0') ? 1 : 0; break;
            case 1: current = (c == '0') ? 2 : 0; break;
            case 2: current = (c == '1') ? 3 : ((c == '0') ? 2 : 0); break;
            case 3: current = 3; break;
        }
    }
    return current;
}

int main() {
    string w;
    cout << "Enter string: ";
    cin >> w;

    int current = NFA_substring(w);
    if (current == 3)
        cout << w << " is accepted" << endl;
    else
        cout << w << " is rejected" << endl;
    return 0;
}

File: sem iv/Theory of Computaion/lab/code/python/9_nfa_contains_001.py
# NFA (simulated deterministically) accepting all strings containing 001 as substring
# States: A=0 (start), B=1 (seen 0), C=2 (seen 00), D=3 (seen 001 -> accepting)

def nfa_substring(w):
    current = 0
    for c in w:
        if current == 0:
            current = 1 if c == '0' else 0
        elif current == 1:
            current = 2 if c == '0' else 0
        elif current == 2:
            if c == '1':
                current = 3
            elif c == '0':
                current = 2
            else:
                current = 0
        else:  # current == 3
            current = 3
    return current

def main():
    w = input("Enter string: ")
    current = nfa_substring(w)
    if current == 3:
        print(f"{w} is accepted")
    else:
        print(f"{w} is rejected")

if __name__ == "__main__":
    main()
